trio/_threads.py

```diff
@@ -240,37 +315,64 @@ def deliver_worker_fn_result(result: outcome.Outcome[RetT]) -> None:
         limiter.release_on_behalf_of(placeholder)
         raise
 
-    def abort(_: RaiseCancelT) -> trio.lowlevel.Abort:
+    def abort(raise_cancel: RaiseCancelT) -> trio.lowlevel.Abort:
+        # fill so from_thread_check_cancelled can raise
+        cancel_register[0] = raise_cancel
         if cancellable:
+            # empty so report_back_in_trio_thread_fn cannot reschedule
             task_register[0] = None
             return trio.lowlevel.Abort.SUCCEEDED
         else:
             return trio.lowlevel.Abort.FAILED
 
-    # wait_task_rescheduled return value cannot be typed
-    return await trio.lowlevel.wait_task_rescheduled(abort)  # type: ignore[no-any-return]
+    while True:
+        # wait_task_rescheduled return value cannot be typed
+        msg_from_thread: ThreadDone[RetT] | Run[object] | RunSync[
+            object
+        ] = await trio.lowlevel.wait_task_rescheduled(abort)
+        if type(msg_from_thread) is ThreadDone:
```

Comment:
Use isinstance(). It's more idiomatic _and_ faster:
```
% python3.8 -m timeit -s "A = type('A', (), {}); a = A()" "isinstance(a, A)"
5000000 loops, best of 5: 44 nsec per loop
% python3.8 -m timeit -s "A = type('A', (), {}); a = A()" "type(a) is A"
5000000 loops, best of 5: 63.4 nsec per loop
```
(the lead is more marginal on more recent Pythons, but still present in my testing)